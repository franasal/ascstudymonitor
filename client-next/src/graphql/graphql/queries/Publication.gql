query Publication($slug: String!) {
  publication: publicationBySlug(slug: $slug) {
    abstract
    authors {
      firstName
      lastName
    }
    created
    disciplines {
      value
    }
    fileAttached
    id
    keywords {
      value
    }
    slug
    journal {
      value
    }
    title
    websites
    year {
      value
    }
    recommendations(first: 5) {
      score
      publication {
        abstract
        authors {
          firstName
          lastName
        }
        created
        disciplines {
          value
        }
        fileAttached
        id
        keywords {
          value
        }
        slug
        journal {
          value
        }
        title
        websites
        year {
          value
        }
      }
    }
  }
}
