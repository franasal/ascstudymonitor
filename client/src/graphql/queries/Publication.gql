query Publication($slug: String!) {
  publicationBySlug(slug: $slug) {
    abstract
    authors {
      firstName
      lastName
    }
    created
    disciplines {
      value
    }
    fileAttached
    id
    keywords {
      value
    }
    slug
    journal {
      value
    }
    title
    websites
    year {
      value
    }
    recommendations(first: 5) {
      score
      publication {
        id
        slug
        title
      }
    }
  }
}
