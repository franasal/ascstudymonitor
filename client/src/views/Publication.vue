<template>
  <div class="entry">
    <span v-if="!publication">Loading..</span>
    <div class="row" v-if="publication" @click="toggleExpand">
      <div class="chevron-wrapper">
        <icon-publication-chevron :expanded="expanded" />
      </div>

      <div class="content">
        <ul class="entry__disciplines">
          <icon-science />
          <li v-for="d in publication.disciplines" v-bind:key="d">
            <a @click="query('disciplines', e)">{{ d }}</a>
          </li>
        </ul>

        <h3>
          {{ publication.title }}

          <icon-download
            @click="download(publication)"
            v-if="publication.file_attached"
            big="true"
          />
        </h3>

        <ul class="entry__authors">
          <icon-author />
          <div class="entry_authors_holder">
            <li
              v-for="authorName in publication.authorNames"
              v-bind:key="authorName"
            >
              <a @click="query('authors', authorName)">{{ authorName }}</a>
            </li>
          </div>
        </ul>
        <slide-up-down :active="expanded" :duration="200">
          <div class="entry__abstract">
            <div class="entry__abstract_inner">
              <icon-abstract />
              <div class="entry__abstract_text" v-if="publication.abstract">
                {{ publication.abstract }}
              </div>
              <div class="entry__abstract_text" v-else>
                Abstract missing.
              </div>
            </div>
          </div>
        </slide-up-down>

        <div class="entry__year_source">
          <a
            class="entry__year_source__year"
            @click="query('year', publixation.year)"
            >{{ publication.year }}</a
          >
          <a @click="query('source', publication.source)">{{
            publication.source
          }}</a>
        </div>

        <div v-if="expanded">
          <div
            class="entry__downloads-item"
            v-for="website in publication.websites"
            v-bind:key="website"
          >
            <icon-link />
            <a target="_blank" rel="noopener noreferrer" :href="website"
              >Visit publisher website
            </a>
          </div>

          <div class="entry__downloads-item" v-if="publication.file_attached">
            <icon-download big="false" />
            <a
              target="_blank"
              rel="noopener noreferrer"
              :href="'/' + publication.id + '/download/'"
              >Download full text</a
            >
          </div>
        </div>
        <!-- <router-link :to="{ path: '/' }">Back to all</router-link> -->
      </div>
    </div>
  </div>
</template>

<script>
// @ is an alias to /src
import SlideUpDown from 'vue-slide-up-down'
import IconDownload from '@/components/Icons/IconDownload.vue'
import IconAuthor from '@/components/Icons/IconAuthor.vue'
import IconLink from '@/components/Icons/IconLink.vue'
import IconAbstract from '@/components/Icons/IconAbstract.vue'
import IconScience from '@/components/Icons/IconScience.vue'
import IconPublicationChevron from '@/components/Icons/IconPublicationChevron.vue'

export default {
  name: 'Publication',
  components: {
    IconDownload,
    IconLink,
    IconAuthor,
    IconAbstract,
    IconScience,
    IconPublicationChevron,
    SlideUpDown,
  },
  data: () => ({
    expanded: false,
  }),
  props: ['slug', 'publicationId'],
  computed: {
    isDetailView() {
      return (
        this.publication &&
        this.$store.state.route.params?.slug === this.publication.slug
      )
    },
    publication() {
      if (this.publicationId)
        return this.$store.state.publications.find(
          p => p.id === this.publicationId,
        )
      else return this.$store.state.publications.find(p => p.slug === this.slug)
    },
  },
  methods: {
    toggleExpand() {
      if (this.isDetailView) return false
      this.expanded = !this.expanded
    },
  },
  created() {
    if (this.isDetailView) {
      this.expanded = true
    }
  },
}
</script>

<style scoped lang="less">
.row {
  display: flex;
}

.chevron-wrapper {
  min-width: 110px;
  display: flex;
  justify-content: center;
  padding-top: 50px;
}

.entry {
  margin: 0;
  padding: 12px 24px 12px 0;
  width: 100%;

  h3 {
    margin-top: 0;
    font-size: 1.8em;
    color: #333;
    font-weight: 700;
    letter-spacing: 0.015em;
    margin-bottom: 10px;
    cursor: pointer;
  }

  ul {
    margin: 0;
  }

  li {
    display: inline-block;
    font-weight: 700;
  }
}

.entry__chevron-wrapper svg {
  transition: transform 0.2s ease-in;
}

.entry__icon {
  width: 35px;
  min-width: 35px;
  display: inline-block;

  .l {
    fill: none;
    stroke: #333;
    stroke-width: 1px;
  }
}

.entry__disciplines {
  list-style: none;
  padding-bottom: 10px;

  a {
    color: #34557f;
    font-size: 1.2em;
    font-weight: 700;
    margin-right: 8px;

    &:hover {
      color: #607a9b;
    }
  }
}

.entry__authors {
  list-style: none;
  padding-bottom: 10px;
  line-height: 2;
  display: inline-flex;

  a {
    color: #333;
    font-weight: 700;
    font-size: 1em;
    margin-right: 8px;

    &:hover {
      color: #607a9b;
    }
  }

  svg {
    height: 18px;
    position: relative;
    bottom: -4px;
    left: 4px;

    .a3 {
      fill: none;
      stroke: #333;
      stroke-width: 1px;
    }
  }
}

.entry__abstract {
  color: #333;
  line-height: 1.5;
  margin-bottom: 12px;
}

.entry__abstract_inner {
  display: inline-flex;
}

.entry__abstract svg {
  margin-left: 3px;
  margin-bottom: -1px;
}

.entry__year_source {
  font-size: 0.9em;
  color: #333;
  opacity: 0.75;
  margin-bottom: 12px;
}

.entry__year_source a {
  display: inline-block;
  color: #333 !important;
}

.entry__year_source__year {
  width: 33px;
}

.entry__year_source a:hover {
  color: #607a9b;
}

.entry__downloads {
  display: inline-flex;
}

/*.entry__downloads-item .entry__icon svg{
  width: 30px;
  min-width: 30px;
}*/

.entry__downloads-item {
  margin-right: 24px;
  display: inline-flex;
  flex-direction: row;
  align-items: center;
  justify-content: stretch;
}

.entry__downloads-item a {
  color: #000;
  font-weight: 600;
}

.entry__downloads-item a:hover {
  color: #607a9b;
  text-decoration: none;
}

.entry__downloads-item a:hover {
  text-decoration: none;
  color: #607a9b;
}

.content {
  padding: 8px 24px 12px 0px;
}
</style>
