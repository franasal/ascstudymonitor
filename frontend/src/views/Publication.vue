<template>
  <div class="entry">
    <span v-if="!publication">Loading..</span>
    <div v-if="publication">
      <ul class="entry__disciplines">
        <icon-science />
        <li v-for="d in publication.disciplines" v-bind:key="d">
          <a @click="query('disciplines', e)">{{ d }}</a>
        </li>
      </ul>

      <router-link :to="{ path: '/publication/' + publication.id }">
        <h3>
          {{ publication.title }}

          <icon-download v-if="publication.file_attached" />
        </h3>
      </router-link>

      <ul class="entry__authors">
        <icon-author />
        <div class="entry_authors_holder">
          <li
            v-for="authorName in publication.authorNames"
            v-bind:key="authorName"
          >
            <a @click="query('authors', authorName)">{{ authorName }}</a>
          </li>
        </div>
      </ul>
      <div class="entry__abstract">
        <div class="entry__abstract_inner">
          <icon-abstract />
          <div class="entry__abstract_text" v-if="publication.abstract">
            {{ publication.abstract }}
          </div>
          <div class="entry__abstract_text" v-else>
            Abstract missing.
          </div>
        </div>
      </div>

      <div class="entry__year_source">
        <a
          class="entry__year_source__year"
          @click="query('year', publixation.year)"
          >{{ publication.year }}</a
        >
        <a @click="query('source', publication.source)">{{
          publication.source
        }}</a>
      </div>

      <div class="entry__downloads" style="display: none">
        <div
          class="entry__downloads-item"
          v-for="website in publication.websites"
          v-bind:key="website"
        >
          <icon-link />
          <a target="_blank" rel="noopener noreferrer" :href="website"
            >Visit publisher website</a
          >
        </div>

        <icon-download v-if="publication.file_attached" />

        <a
          target="_blank"
          rel="noopener noreferrer"
          :href="'/download/' + publication.id"
          >Download full text</a
        >
      </div>
      <router-link :to="{ path: '/' }">Back to all</router-link>
    </div>
  </div>
</template>

<script>
// @ is an alias to /src
import IconDownload from '@/components/IconDownload.vue'
import IconAuthor from '@/components/IconAuthor.vue'
import IconLink from '@/components/IconLink.vue'
import IconAbstract from '@/components/IconAbstract.vue'
import IconScience from '@/components/IconScience.vue'

export default {
  name: 'Publication',
  components: {
    // HelloWorld,
    IconDownload,
    IconLink,
    IconAuthor,
    IconAbstract,
    IconScience,
  },
  props: ['publicationId'],
  computed: {
    publication() {
      console.log('A')
      const p = this.$store.state.publications.find(
        p => p.id === this.publicationId,
      )
      console.log('rendering ,publication is ', p)
      return p
    },
  },
}
</script>

<style>
.entry {
  margin: 0;
  padding: 12px 24px 12px 0;
}

.entry__chevron-wrapper svg {
  transition: transform 0.2s ease-in;
}

.entry ul {
  margin: 0;
}
.entry li {
  display: inline-block;
  font-weight: 700;
}

.entry__icon {
  width: 35px;
  min-width: 35px;
  display: inline-block;
}

.entry__icon .l {
  fill: none;
  stroke: #333;
  stroke-width: 1px;
}

.entry__disciplines {
  list-style: none;
  padding-bottom: 10px;
}
.entry__disciplines a {
  color: #34557f;
  font-size: 1.2em;
  font-weight: 700;
  margin-right: 8px;
}

.entry__disciplines a:hover {
  color: #607a9b;
}

.entry__disciplines .entry__icon {
  width: 32px;
  min-width: 32px;
}

.entry__disciplines svg {
  height: 20px;
  position: relative;
  bottom: -5px;
  left: 0.05em;
}

.entry__disciplines svg .a4 {
  fill: none;
  stroke: #34557f;
  stroke-width: 1px;
}

.entry__authors {
  list-style: none;
  padding-bottom: 10px;
  line-height: 2;
  display: inline-flex;
}

.entry__authors svg {
  height: 18px;
  position: relative;
  bottom: -4px;
  left: 4px;
}

.entry__authors svg .a3 {
  fill: none;
  stroke: #333;
  stroke-width: 1px;
}

.entry__authors a {
  color: #333;
  font-weight: 700;
  font-size: 1em;
  margin-right: 8px;
}

.entry__authors a:hover {
  color: #607a9b;
}

.entry__abstract {
  display: none;
  color: #333;
  line-height: 1.5;
  margin-bottom: 12px;
}

.entry__abstract_inner {
  display: inline-flex;
}

.entry__abstract svg {
  margin-left: 3px;
  margin-bottom: -1px;
}

.entry__year_source {
  font-size: 0.9em;
  color: #333;
  opacity: 0.75;
  margin-bottom: 12px;
}

.entry__year_source a {
  display: inline-block;
  color: #333 !important;
}

.entry__year_source__year {
  width: 33px;
}

.entry__year_source a:hover {
  color: #607a9b;
}

.entry__downloads {
  display: inline-flex;
}

/*.entry__downloads-item .entry__icon svg{
  width: 30px;
  min-width: 30px;
}*/

.entry__downloads-item a {
  color: #000;
  font-weight: 600;
}

.entry__downloads-item a:hover {
  color: #607a9b;
  text-decoration: none;
}

.entry__downloads-item a:hover {
  text-decoration: none;
  color: #607a9b;
}

.entry__downloads-icon {
  height: 24px;
  width: 24px;
  border-radius: 24px;
  display: flex;
  align-items: center;
  justify-content: center;
  margin-right: 10px;
}

.entry__downloads-icon__blue {
  background-color: rgb(27, 157, 164);
}

.entry__downloads-icon svg {
  height: 14px;
  position: relative;
  right: -0.03em;
  top: -0.02em;
}

.entry__downloads-item.big {
  display: inline-block;
  height: 24px;
  width: 24px;
  border-radius: 24px;
  /*margin-right: 10px;*/
}

.entry__downloads-item.big svg {
  height: 14px;
  position: relative;
  right: -0.01em;
  top: -0.02em;
}
</style>
