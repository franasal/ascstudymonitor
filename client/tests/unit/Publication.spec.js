import { shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import Publication from '@/views/Publication.vue'

const localVue = createLocalVue()
localVue.use(Vuex)

describe('Publication.vue', () => {
  let store

  beforeEach(() => {
    store = new Vuex.Store({
      state: {
        route: {},
        publications: [
          {
            id: 'A',
            slug: '123456-A',
            websites: ['http://test.com'],
          },
        ],
      },
    })
  })

  it('matches snapshot', () => {
    const wrapper = shallowMount(Publication, { store, localVue })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('renders the entry', () => {
    const wrapper = shallowMount(Publication, { store, localVue })
    expect(wrapper.find('.entry').exists()).toBe(true)
  })

  it('expands', async () => {
    const wrapper = shallowMount(Publication, {
      store,
      localVue,
      propsData: { publicationId: 'A' },
    })

    let downloads = wrapper.find('.entry__downloads-item')
    expect(downloads.element).toBeFalsy()

    wrapper.find('.chevron-wrapper').trigger('click')

    await wrapper.vm.$nextTick()

    downloads = wrapper.find('.entry__downloads-item')

    expect(downloads.isVisible()).toBe(true)
  })
})
