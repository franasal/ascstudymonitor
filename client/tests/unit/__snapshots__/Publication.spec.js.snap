// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Publication.vue matches snapshot 1`] = `
<div
  class="entry"
>
  <span>
    Loading..
  </span>
   
  <!---->
</div>
`;
